core/station/impl/src/services/system.rs

```diff
@@ -575,69 +662,257 @@ mod init_canister_sync_handlers {
         }
     }
 
+    pub async fn set_initial_user_groups(
+        user_groups: &[InitUserGroupInput],
+    ) -> Result<(), ApiError> {
+        let add_user_groups = user_groups
+            .iter()
+            .map(|user_group| {
+                let input = AddUserGroupOperationInput {
+                    name: user_group.name.clone(),
+                };
+
+                (
+                    input,
+                    *HelperMapper::to_uuid(user_group.id.clone())
+                        .expect("Invalid UUID")
+                        .as_bytes(),
+                )
+            })
+            .collect::<Vec<(AddUserGroupOperationInput, UUID)>>();
+
+        for (new_user_group, with_user_group_id) in add_user_groups {
+            USER_GROUP_SERVICE
+                .create_with_id(new_user_group, Some(with_user_group_id))
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    pub fn set_initial_named_rules(named_rules: &[InitNamedRuleInput]) -> Result<(), ApiError> {
+        let mut add_named_rules = named_rules
+            .iter()
+            .map(|named_rule| {
+                let input = AddNamedRuleOperationInput {
+                    name: named_rule.name.clone(),
+                    description: named_rule.description.clone(),
+                    rule: named_rule.rule.clone().into(),
+                };
+
+                (
+                    input,
+                    HelperMapper::to_uuid(named_rule.id.clone()).map(|uuid| *uuid.as_bytes()),
+                )
+            })
+            .map(|(input, result)| result.map(|uuid| (input, uuid)))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // sorting criteria:
+        // - if a policy depends on another policy, the dependent policy should be added first
+        // - keep the original order of the policys otherwise
+        add_named_rules.sort_by(|a, b| {
+            if b.0.rule.has_named_rule_id(&a.1) {
+                Ordering::Less
+            } else {
+                Ordering::Greater
```

Comment:
I'm not sure this result leads to keeping the original order: shouldn't we compare the original indices here?

Reply:
replacing `Greater` with `Equal`, since the sort is stable, it should keep the original order for independent named rules/policies